lib/offline.js

```diff
@@ -58,14 +61,48 @@ module.exports = promisify(function(config, callback) {
       return '';
     }
   }).then(function(cacheId) {
-    swPrecache.write(path.join(rootDir, 'offline-worker.js'), {
-      staticFileGlobs: fileGlobs.map(function(v) { return path.join(rootDir, v) }),
-      stripPrefix: rootDir,
-      verbose: true,
-      logger: gutil.log,
-      importScripts: config.importScripts || [],
-      cacheId: cacheId,
-      ignoreUrlParametersMatching: config.ignoreUrlParametersMatching || [/./],
-    }, callback);
+    var absoluteGlobs =
+      fileGlobs.map(function (v) { return path.join(rootDir, v); });
+    var files = flatGlobs(absoluteGlobs);
+    var filesAndHashes = getFilesAndHashes(files, rootDir);
+    var replacements = { resources: filesAndHashes, name: cacheId };
+
+    var stream = gulp.src([__dirname + '/../templates/app/offline-worker.js'])
+      .pipe(template(replacements))
+      .pipe(conflict(rootDir))
+      .pipe(gulp.dest(rootDir));
+
+    stream.on('finish', function () { callback(); });
+    stream.on('error', function (e) { callback(e); });
   });
+
+  function flatGlobs(fileGlobs) {
+    return Object.keys(fileGlobs.reduce(function (matchings, fileGlob) {
+      fileGlob = fileGlob.replace(path.sep, '/');
+      glob.sync(fileGlob, { nodir: true }).forEach(function (m) {
+        matchings[m] = m;
+      });
+      return matchings;
+    }, {}));
+  }
+
+  function getFilesAndHashes(files, stripPrefix) {
+    return files
+      .map(function (filepath) {
+        var data = fs.readFileSync(filepath);
+        var hash = getHash(data);
+        return { filepath: filepath.replace(stripPrefix, ''), hash: hash };
```

Comment:
This will strip the prefix elsewhere in the string, if it doesn't appear at the beginning. That's unlikely, but in the event that it happens, we should avoid it, which you could do with a function second parameter that checks the offset of the string to replace:

```
filepath.replace(stripPrefix, function(v, offset) { if (offset === 0) return ''; });
```

It might also make sense to warn if the prefix isn't at the beginning of the path, unsure.